### PR TITLE
Add logger runtime trace level change possibility

### PR DESCRIPTION
**Pull request by ktrzcinx**
Why:
To improve debugging process.

How:
Config file with list of new trace level for particular UUID components will be translated and send to FW in runtime (logger->driver->IPC->FW). In FW current log level threshold should be overwritten with value from IPC message. Config file is given as logger runtime argument.

Implements:
https://github.com/thesofproject/sof/issues/2172

### CODE REVIEW

**Comment by dbaluta** [comment]
Cc @crazoes this is close related with what we want to do for summer. Lets keep an eye on this. Please try to review and understand this patch.

File: src/include/sof/trace/trace.h [hunk 1]
```diff
@@ -305,7 +305,7 @@ struct tr_ctx {
  * @param default_log_level Default log level.
  */
```
Comment (dbaluta) [inline src/include/sof/trace/trace.h:309]: @ktrzcinx Please be more generous with the explanations in the commit message. 

It is not very clear which entries will be more easier to iterate over. Also, is there a typo here? 'over entries of ..'  ->  'over entries OR'

Comment (paulstelian97) [reply src/include/sof/trace/trace.h:309]: Wording seems a bit obtuse. I'd rephrase to "It is the first step to allow iterating over the UUID (or trace context?) entries and searching for the right one."

File: src/audio/buffer.c [hunk 2]
```diff
@@ -32,31 +32,32 @@ struct comp_buffer *buffer_alloc(uint32_t size, uint32_t caps, uint32_t align)
 
 	/* validate request */
 	if (size == 0 || size > HEAP_BUFFER_SIZE) {
-		buf_cl_err("buffer_alloc(): new size = %u is invalid", size);
+		tr_err(&buffer_tr, "buffer_alloc(): new size = %u is invalid",
+		       size);
```
Comment (dbaluta) [inline src/audio/buffer.c:36]: So, can we remove buf_cl_err?

File: src/include/ipc/trace.h [hunk 3]
```diff
@@ -51,6 +51,19 @@ struct sof_ipc_dma_trace_posn {
 	uint32_t messages;	/* total trace messages */
 } __attribute__((packed));
 
+/* part of sof_ipc_dma_trace_filter */
+struct sof_ipc_dma_trace_filter_elem {
+	uint32_t uuid_addr;	/* component type identification */
+	int32_t lvl;		/* log level */
+} __attribute__((packed));
```
Comment (dbaluta) [inline src/include/ipc/trace.h:58]: Just log_level here. lvl is pretty hard to read.

File: src/include/ipc/header.h [hunk 4]
```diff
@@ -152,6 +152,7 @@
 #define SOF_IPC_TRACE_DMA_PARAMS		SOF_CMD_TYPE(0x001)
```
Comment (dbaluta) [inline src/include/ipc/header.h:152]: @ktrzcinx any idea who sets the initial trace level?

File: src/trace/trace.c [hunk 5]
```diff
@@ -113,9 +113,22 @@ bool trace_filter_update(uint32_t uuid_addr, int trace_level)
 	return -EINVAL;
 }
 
+/**
+ * \brief Runtime trace filtration
+ * \param lvl log level (error, info, debug ...)
+ * \param uuid uid address
```
Comment (dbaluta) [inline src/trace/trace.c:119]: filtering would be more be appropriate here

File: src/trace/trace.c [hunk 6]
```diff
@@ -113,9 +113,22 @@ bool trace_filter_update(uint32_t uuid_addr, int trace_level)
...
+ * \return false when trace is filtered out, otherwise true
+ */
+static inline bool trace_filter_pass(uint32_t lvl,
+				     const struct tr_ctx *ctx)
+{
+	/* LOG_LEVEL_CRITICAL has low value, LOG_LEVEL_VERBOSE big */
+	return lvl <= ctx->level;
```
Comment (dbaluta) [inline src/trace/trace.c:126]: Lets go with high instead of big here.

**Comment by dbaluta** [comment]
@ktrzcinx @lgirdwood can someone start writing some documentation for logging in SOF. 
@crazoes @IulianOlaru249 will jump with questions and review. 

As far as I figured out there are two ways of moving the logs from DSP to kernel:

1) shared memory (mbox)
2) DMA trace

I assume 2) was added because 1) has very small place to hold the logs, right?

File: tools/logger/logger.c [hunk 7]
```diff
@@ -141,6 +143,7 @@ static int configure_uart(const char *file, unsigned int baud)
 
 int main(int argc, char *argv[])
 {
+	static const char opt_frm[] = "ho:i:l:ps:c:u:tev:rdLf:F:g:";
 	struct convert_config config;
```
Comment (dbaluta) [inline tools/logger/logger.c:147]: Please introduce opt_frm in a separate patch? And why is it static?

File: tools/logger/convert.h [hunk 8]
```diff
@@ -47,4 +48,6 @@ struct convert_config {
 void log_err(FILE *out_fd, const char *fmt, ...);
 uint32_t get_uuid_key(const struct snd_sof_uids_header *uids_dict,
 		      const struct sof_uuid_entry *entry);
+char *format_uid_raw(const struct sof_uuid_entry *uid_entry, int use_colors,
+		     int name_first);
 int convert(struct convert_config *config);
```
Comment (dbaluta) [inline tools/logger/convert.h:53]: Where is the implementation of format_uid_raw? If the implementation is another patch of this series then add this prototype to that patch.

If format_uid_raw already exists in the code, then add the prototype in a separate patch.

File: tools/logger/logger.c (hunk 7 above)
Comment (dbaluta) [inline tools/logger/logger.c:147]: Lets name opt_frm -> optstring as the man page of getopt(3) names it.

File: src/include/ipc/header.h (hunk 4 above)
Comment (ktrzcinx) [reply src/include/ipc/header.h:152]: `DECLARE_TR_CTX(buffer_tr, SOF_UUID(buffer_uuid), LOG_LEVEL_INFO);`

File: src/trace/trace.c [hunk 9]
```diff
@@ -96,6 +96,23 @@ static inline void mtrace_event(const char *data, uint32_t length)
...
+	struct tr_ctx *end = (struct tr_ctx *)&_trace_ctx_end;
+
+	while (ptr < end) {
+		if (ptr->uuid_p == uuid_addr) {
+			ptr->level = trace_level;
+			return 0;
+		}
+		++ptr;
+	}
+	return -EINVAL;
```
Comment (akloniex) [inline src/trace/trace.c:113]: Function's return type is bool, so -EINVAL isn't a proper return value.

File: tools/logger/filter.c [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+	if (!out_fd) {
+		log_err(NULL, "unable to open %s file, %d",
+			FILTER_KERNEL_PATH, errno);
+		ret = -errno;
+		goto err;
+	}
+
+	line = malloc(128);
+	if (!line)
+		return -ENOMEM;
```
Comment (akloniex) [inline tools/logger/filter.c:96]:
```
{
ret = -ENOMEM;
goto err;
}
```

File: src/audio/buffer.c (hunk 2 above)
Comment (ktrzcinx) [reply src/audio/buffer.c:36]: removed

File: src/include/sof/trace/trace.h (hunk 1 above)
Comment (ktrzcinx) [reply src/include/sof/trace/trace.h:309]: updated

File: src/include/ipc/trace.h (hunk 3 above)
Comment (ktrzcinx) [reply src/include/ipc/trace.h:58]: changed

File: src/trace/trace.c (hunk 5 above)
Comment (ktrzcinx) [reply src/trace/trace.c:119]: changed

File: src/trace/trace.c (hunk 6 above)
Comment (ktrzcinx) [reply src/trace/trace.c:126]: changed

File: tools/logger/logger.c (hunk 7 above)
Comment (ktrzcinx) [reply tools/logger/logger.c:147]: I can't see any profit after removing `static` keyword

File: tools/logger/convert.h (hunk 8 above)
Comment (ktrzcinx) [reply tools/logger/convert.h:53]: This function is already implemented in converter.c

File: tools/logger/logger.c (hunk 7 above)
Comment (ktrzcinx) [reply tools/logger/logger.c:147]: changed

File: src/trace/trace.c (hunk 9 above)
Comment (ktrzcinx) [reply src/trace/trace.c:113]: changed return type to int

File: tools/logger/filter.c (hunk 10 above)
Comment (ktrzcinx) [reply tools/logger/filter.c]: right

File: tools/logger/logger.c (hunk 7 above)
Comment (dbaluta) [reply tools/logger/logger.c:147]: The profit will be the code clarity. Static variables are rarely used inside functions and when used they have a special purpose.

Comment (ktrzcinx) [reply tools/logger/logger.c:147]: `static const` is in general better approach, because then variable is initialized only once - in opposite to `const` which will be initialized in each function call. For `main` function there is actually no difference, but we should keep good practice. Moreover after removing `static` checkpatch complain: `WARNING: const array should probably be static const`

File: src/ipc/handler.c [hunk 11]
```diff
@@ -738,6 +738,30 @@ static int ipc_dma_trace_config(uint32_t header)
...
+	tr_info(&ipc_tr, "ipc: trace_filter_update received, size %d elems",
+		packet->elem_cnt);
+
+	for (i = 0; i < packet->elem_cnt; ++i) {
+		elem = &packet->elems[i];
+		ret = trace_filter_update(elem->uuid_key, elem->log_level);
+		if (ret) {
+			tr_info(&ipc_tr, "ipc: trace_filter_update error for UUID key 0x%X and log level %d",
+				elem->uuid_key, elem->log_level);
+			return ret;
```
Comment (lgirdwood) [inline src/ipc/handler.c:758]: This should really keep going, but return the error status at the end. This way at least the user has most of the trace setting updated. Please also add comments to this functions.

File: src/trace/trace.c [hunk 12]
```diff
@@ -96,6 +96,23 @@ static inline void mtrace_event(const char *data, uint32_t length)
...
 #endif /* CONFIG_TRACEM */
 
+int trace_filter_update(uint32_t uuid_addr, int trace_level)
+{
+	extern void *_trace_ctx_start;
+	extern void *_trace_ctx_end;
+	struct tr_ctx *ptr = (struct tr_ctx *)&_trace_ctx_start;
+	struct tr_ctx *end = (struct tr_ctx *)&_trace_ctx_end;
+
+	while (ptr < end) {
```
Comment (lgirdwood) [inline src/trace/trace.c:106]: What are we doing here, please add comments.

File: src/trace/trace.c [hunk 13]
```diff
@@ -113,9 +113,22 @@ int trace_filter_update(uint32_t uuid_addr, int trace_level)
...
+/**
+ * \brief Runtime trace filtering
+ * \param lvl log level (LOG_LEVEL_ ERROR, INFO, DEBUG ...)
+ * \param uuid uid address
+ * \return false when trace is filtered out, otherwise true
+ */
+static inline bool trace_filter_pass(uint32_t lvl,
+				     const struct tr_ctx *ctx)
+{
+	/* LOG_LEVEL_CRITICAL has low value, LOG_LEVEL_VERBOSE high */
```
Comment (lgirdwood) [inline src/trace/trace.c:125]: Critical/panic should never be filtered, can you confirm.

File: tools/logger/convert.c [hunk 14]
```diff
@@ -141,6 +141,13 @@ get_uuid_entry(const struct snd_sof_uids_header *uids_dict, uint32_t uid_ptr)
 	       uids_dict->base_address);
 }
 
+uint32_t get_uuid_key(const struct snd_sof_uids_header *uids_dict,
```
Comment (lgirdwood) [inline tools/logger/convert.c:145]: static inline ?

File: tools/logger/convert.c [hunk 15]
```diff
@@ -141,6 +141,13 @@ get_uuid_entry(const struct snd_sof_uids_header *uids_dict, uint32_t uid_ptr)
...
+		      const struct sof_uuid_entry *entry)
+{
+	return (uintptr_t)entry - (uintptr_t)uids_dict -
```
Comment (lgirdwood) [inline tools/logger/convert.c:148]: What does this do, can you describe how the dictionary format and how the look works.

File: src/include/user/abi_dbg.h [hunk 16]
```diff
@@ -19,7 +19,7 @@
 #define __USER_ABI_DBG_H__
 
 #define SOF_ABI_DBG_MAJOR 5
-#define SOF_ABI_DBG_MINOR 0
+#define SOF_ABI_DBG_MINOR 1
```
Comment (lgirdwood) [inline src/include/user/abi_dbg.h:22]: Kernel is not translating UUIDs, it passing them directly for topology loading.

File: tools/logger/filter.c [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+
+	fprintf(out_fd, "\n");
+
+err:
+	if (config_fd)
+		fclose(config_fd);
+	if (out_fd)
+		fclose(out_fd);
+	return ret;
+}
```
Comment (lgirdwood) [inline tools/logger/filter.c:132]: This all needs comments. Every block. What does it do...

File: tools/logger/CMakeLists.txt [hunk 18]
```diff
@@ -3,6 +3,7 @@
 add_executable(sof-logger
 	logger.c
 	convert.c
+	filter.c
```
Comment (lgirdwood) [inline tools/logger/CMakeLists.txt:6]: So my understanding is

1) create UUID filter text file.
2) pass UUID fillet to logger ?

This seems overly complex and cumbersome. UUID is also not the instance ID, meaning we enable all UUIDs for typeX and not just the one we care about

We should really be filtering on component instance name. e.g. BUF1.1, VOL2..0 etc and do it via the command line
```sof-logger -fdebug="BUF1.1, VOL1.0" -finfo="BASEFW" -fvdebug="SRC1.1"```

We should also be able to select the pipeline too

```sof-logger -fdebug="PIPELINE1"```

File: src/audio/buffer.c [hunk 19]
```diff
@@ -22,41 +22,42 @@
 /* 42544c92-8e92-4e41-b679-34519f1c1d28 */
 DECLARE_SOF_UUID("buffer", buffer_uuid, 0x42544c92, 0x8e92, 0x4e41,
 		 0xb6, 0x79, 0x34, 0x51, 0x9f, 0x1c, 0x1d, 0x28);
-uintptr_t buffer_uuid_ptr = SOF_UUID(buffer_uuid);
+DECLARE_TR_CTX(buffer_tr, SOF_UUID(buffer_uuid), LOG_LEVEL_INFO);
```
[the remark this answers is not in the document]
Comment (ktrzcinx) [reply src/audio/buffer.c:25]:
```
/**
 * Trace context.
 */
struct tr_ctx {
	uintptr_t uuid_p;	/**< UUID pointer, use SOF_UUID() to init */
	uint32_t level;		/**< Default log level */
};
```

`LOG_LEVEL_INFO` is default log level for `buffer_tr` (default log threshold). This value may be changed in runtime, to change current trace level for each log function referencing to `buffer_tr`. So it's related more with tracing component rather than trace log message.
On the other side we have log level, related with log message. This one is defined in `tr_info`, `tr_err`... functions and it can't be change in runtime.

File: src/include/sof/audio/buffer.h [hunk 20]
```diff
@@ -33,45 +33,25 @@
...
-			     -1, -1, __e, ##__VA_ARGS__)
-
-#define buf_cl_dbg(__e, ...) \
-	tracev_event_with_ids(TRACE_CLASS_DEPRECATED, buffer_uuid_ptr, \
-			      -1, -1, __e, ##__VA_ARGS__)
+extern struct tr_ctx buffer_tr;
 
 #define buf_err(buf_ptr, __e, ...) \
-	trace_error_with_ids(TRACE_CLASS_DEPRECATED, buffer_uuid_ptr, \
+	trace_error_with_ids(TRACE_CLASS_DEPRECATED, &buffer_tr, \
```
[the remark this answers is not in the document]
Comment (ktrzcinx) [reply src/include/sof/audio/buffer.h:39]: perhaps I should do it in this commit ;)

File: src/include/user/abi_dbg.h (hunk 16 above)
Comment (ktrzcinx) [reply src/include/user/abi_dbg.h:22]: In proposed solution, it is as you wrote. Here logger is responsible for translation UUID value to UUID key/address value.

File: tools/logger/CMakeLists.txt (hunk 18 above)
Comment (ktrzcinx) [reply tools/logger/CMakeLists.txt:6]: What's a use case for such a trace logger granuality? I guess it's rather error prone solution, because of trace filtering parameters incompatibility between different platforms/topologies.

**Comment by mmaka1** [comment]
> One question: In addition to this runtime configurability, is it still possible to use Kconfig to give default values for the logging levels? Or do we need to add per-component configs for this in the future?

@paulstelian97 Each trace context's level is initialized to a default value (LOG_LEVEL_INFO atm) with a DECLARE_TR_CTX() parameter. This may be easily replaced with a new CONFIG_DEFAULT_LOG_LEVEL (set to info) defined by the Kconfig and overridden in selected places with another Kconfig-defined default (set to error for instance) if needed.

File: tools/logger/CMakeLists.txt (hunk 18 above)
Comment (ktrzcinx) [reply tools/logger/CMakeLists.txt:6]: Also when we would like to have verbose logs from many components, then cmd line to run might be quite long and unreadable. Moreover file usage, guarantee the same configuration between debug sessions, what may be not true for command line arguments.

Comment (paulstelian97) [reply tools/logger/CMakeLists.txt:6]: Can't unrecognized parameters simply be ignored, perhaps with a warning in the traces?

File: tools/logger/convert.c (hunk 15 above)
Comment (paulstelian97) [reply tools/logger/convert.c:148]: The code was moved into a separate function. This arithmetic is required to translate from the runtime location to an address inside the ldc file. uids_dict->data_offset and uids_dict->base_address are both constants, uids_dict is a mmaped region from the ldc file presumably. We are essentially translating from the unusable address with base 0x1B.. whatever to an actual usable address inside the mmaped .ldc file.

File: tools/logger/CMakeLists.txt (hunk 18 above)
Comment (ktrzcinx) [reply tools/logger/CMakeLists.txt:6]: I rather thought about situation when given parameters are valid, but between different topology refer to differ components (in sense of usage, not type), so it may be unintended change.

File: tools/logger/filter.c [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	line = malloc(128);
+	if (!line) {
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	while (1) {
+		read = getline(&line, &len, config_fd);
```
Comment (paulstelian97) [inline tools/logger/filter.c:101]: I'm not a fan of a local variable shadowing a POSIX system call. I'd suggest line_length perhaps? Or ret2 if you don't actually care about the specific value when it's positive.

File: tools/logger/convert.c (hunk 15 above)
Comment (ktrzcinx) [reply tools/logger/convert.c:148]: base is 0x1FFFA000, but rest is right

File: tools/logger/filter.c [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+		if (read < 0)
+			break;
+		if (read < 2 || line[0] == '#')
+			continue;
+		parsed = parse_uuid(line, &uuid);
+		uuid_key = translate_uuid(uids_dict, &uuid);
+		if (parsed < 0 || !uuid_key) {
+			log_err(NULL, "Invalid uuid entry: %s", line);
```
Comment (paulstelian97) [inline tools/logger/filter.c:109]: For compatibility I suggest moving the !uuid_key check lower, have it emit a warning instead of failing and do the fprintf on line 121 (as of this review) only if uuid_key != 0.

File: tools/logger/filter.c [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+	uint32_t c;
+	uint32_t d0;
+	uint64_t d2;
+	int n;
+	int ret;
+
+	ret = sscanf(src, " <%X-%X-%X-%X-%lX>%n", &dst->a, &b, &c, &d0, &d2,
+		     &n);
+	if (ret != 5 || b > UINT32_MAX || c > UINT32_MAX)
+		return -EINVAL;
```
Comment (paulstelian97) [inline tools/logger/filter.c:30]: I suppose clobbering dst->a is fine but clobbering dst->b or dst->c is not?

Also I seriously doubt you can ever get b > UINT32_MAX given that b is a uint32_t.

File: src/include/user/abi_dbg.h (hunk 16 above)
Comment (lgirdwood) [reply src/include/user/abi_dbg.h:22]: The commit message is saying the kernel will do this (and there is no way to comment on the commit message).

File: tools/logger/CMakeLists.txt (hunk 18 above)
Comment (lgirdwood) [reply tools/logger/CMakeLists.txt:6]: @ktrzcinx use case is that I want to trace pipeline 0 (comps X, Y) , and not the same component type on other (connected or otherwise) pipelines. I can also read the topology (or view the topology image) and know the strings I need to pass without have to manually convert topology elements to UUIDs

Please also state why you think it's error prone ? Copying a 32 byte ASCII is UUID is error prone.

Command line usage is exactly the same as UUID file usage. It can be scripted or sent via email if needed.

File: tools/logger/filter.c (hunk 21 above)
Comment (dbaluta) [reply tools/logger/filter.c:101]: Or just simply len.

File: tools/logger/CMakeLists.txt (hunk 18 above)
Comment (ktrzcinx) [reply tools/logger/CMakeLists.txt:6]: I think about such a situation - we have topology 1 like:
`BUF1.1 -> VOL1.1 -> BUF1.2 -> SmartAmp1.2 -> BUF1.3 -> DAI1.3`
`BUF2.1 -> FIR2.1 -> BUF2.2 -> SmartAmp2.2 -> BUF2.3 -> DAI2.3`
and we found some issue on SmartAmp2.2, what is seen only in verbose logs for SmartAmp connected after FIR component what isn't known yet (hypothetical situation). After try of reproduction on another platform we didn't see any troubles, because topology 2 is:
`BUF1.1 -> FIR1.1 -> BUF1.2 -> SmartAmp1.2 -> BUF1.3 -> DAI1.3`
`BUF2.1 -> VOL2.1 -> BUF2.2 -> SmartAmp2.2 -> BUF2.3 -> DAI2.3`

*) Component connections layout is only for example purpose
  
Here, different tracking levels for different `SmartAmp` hide error, but filtering parameters are valid for both topology. 

32 byte ASCII can be easily validated, readed from source code, output logs, ldc file dump or generated filter template config file. Moreover UUID doesn't change at all, so basically we need to check it once, in opposite to component instance name which one need to be check for each topology individually.

Comment (lgirdwood) [reply tools/logger/CMakeLists.txt:6]: @ktrzcinx sorry not getting your point here. If I want to debug pipeline 1 I want to

```sof-logger -fdebug="pipeline1"```

Thats all. Likewise if I want to debug all smart amps I want to

```sof-logger -fdebug="SmartAmp*"```

My key point is that the debugging must be really simple and intuative. e.g. I want to enable debug for everything

```sof-logger --fdebug="*"```

or I want to debug pipeline2 except for the DAI (as it can spam my logs)

```sof-logger -fdebug="pipeline2, !DAI2.3```

or I want to debug the allocator and FIRs

```sof-logger -fdebug="sys.alloc, FIR*"```

Comment (ktrzcinx) [reply tools/logger/CMakeLists.txt:6]: > `sof-logger -fdebug="pipeline2, !DAI2.3`

Perhaps following this convention, such an example also should be valid:
`sof-logger -fdebug="pipeline2, !DAI2.3, !DAI2.1`
What results in quite complicated messages parsing and handling. When we little change convention, and do it like this:
`sof-logger -fdebug="pipeline2" -ferror="DAI2.3, DAI2.1"`
Sending messages from left to right, so following config can overwrite previous one, then implementation will be much easier to implement and maintain.

Comment (lgirdwood) [reply tools/logger/CMakeLists.txt:6]: @ktrzcinx yep, both examples you gave there would be valid. The last passed argument overrides previous args (this makes the parsing simple). I would imagine you would have a list of all objects (with a trace level variable) and set the level for each object as the args were parsed. Yes, so previous args could be overridden, but that's OK and expected (i.e. GCC does this).
Do you need an IPC from the kernel to get all the valid objects or can you read all the data you need today via sysFS ? If you need a new IPC then it should export all the data you need via sysfs. i.e. reading the sysfs file would send the IPC to FW.
Btw, we would also need to standardise the naming of the basefw "system" components, maybe we could have a "sys." prefix i.e. "sys.allocator", "sys.scheduler.edf" and so on (we can then use "sys.*" as an arg.

**Comment by ktrzcinx** [comment]
Commits before `ipc: trace: Add trace filter message handling` are in separate, linked PR's

File: src/include/sof/trace/trace.h [hunk 24]
```diff
@@ -306,7 +306,7 @@ struct tr_ctx {
...
 #define DECLARE_TR_CTX(ctx_name, uuid, default_log_level) \
-	struct tr_ctx ctx_name = { \
+	struct tr_ctx ctx_name __section(".trace_ctx") = { \
```
Comment (lgirdwood) [inline src/include/sof/trace/trace.h:312]: Can you say why in the commit messge.

File: src/include/ipc/trace.h [hunk 25]
```diff
@@ -51,6 +51,22 @@ struct sof_ipc_dma_trace_posn {
 	uint32_t messages;	/* total trace messages */
 } __attribute__((packed));
 
+/** part of sof_ipc_dma_trace_filter */
+struct sof_ipc_dma_trace_filter_elem {
+	int32_t uuid_id;	/**< type id, or -1 when not important */
```
Comment (lgirdwood) [inline src/include/ipc/trace.h:56]: Better to have 
```c
struct elem {
    enum type type;
    int id;
    int level;
};
```
This way adding new types means a minor ABI update

File: src/ipc/handler.c [hunk 26]
```diff
@@ -738,6 +738,38 @@ static int ipc_dma_trace_config(uint32_t header)
...
+	int ret = 0;
+	int temp;
+	int i;
+
+	tr_info(&ipc_tr, "ipc: trace_filter_update received, size %d elems",
+		packet->elem_cnt);
+
+	packet = ipc->comp_data;
+
+	for (i = 0; i < packet->elem_cnt; ++i) {
```
Comment (lgirdwood) [inline src/ipc/handler.c:755]: inline comment.

File: src/trace/trace.c [hunk 27]
```diff
@@ -96,6 +97,91 @@ static inline void mtrace_event(const char *data, uint32_t length)
...
+	switch (icd->type) {
+	case COMP_TYPE_COMPONENT:
+		return &icd->cd->tctx;
+	case COMP_TYPE_BUFFER:
+		return &icd->cb->tctx;
+	case COMP_TYPE_PIPELINE:
+		return &icd->pipeline->tctx;
+	}
+	/* each COMP_TYPE must be specified */
+	assert(NULL);
```
Comment (lgirdwood) [inline src/trace/trace.c:134]: Is this really fatal ?

File: src/trace/trace.c [hunk 28]
```diff
@@ -96,6 +97,91 @@ static inline void mtrace_event(const char *data, uint32_t length)
...
+					 int32_t comp_id, int32_t log_level)
+{
+	struct ipc *ipc = ipc_get();
+	struct ipc_comp_dev *icd;
+	struct list_item *clist;
+	struct tr_ctx *ctx;
+	bool correct_comp;
+	int cnt = 0;
+
+	list_for_item(clist, &ipc->comp_list) {
```
Comment (lgirdwood) [inline src/trace/trace.c:148]: inline comment

File: tools/logger/convert.c (hunk 15 above)
Comment (lgirdwood) [reply tools/logger/convert.c:152]: ok, can we have this in a comment.

File: tools/logger/logger.c [hunk 29]
```diff
@@ -138,9 +140,36 @@ static int configure_uart(const char *file, unsigned int baud)
 	return ret < 0 ? -errno : fd;
 }
 
+static int append_filter_config(struct convert_config *config,
```
Comment (lgirdwood) [inline tools/logger/logger.c:143]: this needs inline comments.

File: tools/logger/filter.c [hunk 30]
```diff
@@ -0,0 +1,316 @@
...
+#include "filter.h"
+
+#define COMPONENTS_SEPARATOR ','
+
+/** map between log level given by user and enum value */
+static const struct {
+	const char name[16];
+	int32_t log_level;
+} log_level_dict[] = {
+	{"verbose",	LOG_LEVEL_VERBOSE},
```
Comment (lgirdwood) [inline tools/logger/filter.c:28]: vrebose debug ?

File: tools/logger/filter.c [hunk 31]
```diff
@@ -0,0 +1,316 @@
...
+ */
+static int filter_parse_component(const struct snd_sof_uids_header *uids_dict,
+				  char *input_str,
+				  struct sof_ipc_dma_trace_filter_elem *out)
+{
+	struct sof_uuid_entry *uuid_entry;
+	char comp_name[UUID_NAME_MAX_LEN];
+	char scan_format_string[32];
+	const char *instance_info;
+	static const struct sof_uuid sof_pipe_uuid = {
```
Comment (lgirdwood) [inline tools/logger/filter.c:134]: We should not be hard coding these system UUIDs. I think we probably need a shared header ?

**Comment by lgirdwood** [comment]
@ktrzcinx fwiw, UUID now merged. UUID topologies now merging this week so should be good to go soon.

**Comment by lgirdwood** [comment]
@ktrzcinx I think we are good to go now that UUID is all merged.

**Comment by ktrzcinx** [comment]
@lgirdwood great. Now I need alignment component id between topology and value sent in trace's, @fredoh9 is going to implement this part. To speed up the process, we can decouple this alignment and implement filtration mechanism based on component unique id (set in driver in runtime), which one can be seen in output logs ATM. From filtration point of view, changing component unique id to component topology id will be only "one line change", so I guess it will be good direction.

File: tools/logger/filter.c (hunk 30 above)
Comment (ktrzcinx) [reply tools/logger/filter.c:28]: It's just LOG_LEVEL_* enum representation, in user/trace.h file there is 
`#define LOG_LEVEL_VERBOSE	LOG_LEVEL_DEBUG`

File: src/include/ipc/trace.h (hunk 25 above)
Comment (ktrzcinx) [reply src/include/ipc/trace.h]: Updated to similar key-value version.

File: src/ipc/handler.c (hunk 26 above)
Comment (ktrzcinx) [reply src/ipc/handler.c]: Commend added

File: src/trace/trace.c (hunk 28 above)
Comment (ktrzcinx) [reply src/trace/trace.c:203]: added

File: tools/logger/logger.c (hunk 29 above)
Comment (ktrzcinx) [reply tools/logger/logger.c:146]: added

File: src/trace/trace.c (hunk 12 above)
Comment (ktrzcinx) [reply src/trace/trace.c:155]: added

File: tools/logger/convert.c (hunk 15 above)
Comment (ktrzcinx) [reply tools/logger/convert.c:160]: added

File: tools/logger/convert.c (hunk 14 above)
Comment (ktrzcinx) [reply tools/logger/convert.c:151]: This function is strictly related with `get_uuid_entry`, they just work in opposite directions, so I would like to keep them together, but only one of them is used in other file, so I will keep it as it is.

File: tools/logger/filter.c (hunk 31 above)
Comment (ktrzcinx) [reply tools/logger/filter.c]: I have try to move pipe uuid A/B/C/D values to headers in user/ folder, but this solution doesn't look nice. So I just remove this part.

File: src/include/sof/trace/trace.h (hunk 24 above)
Comment (ktrzcinx) [reply src/include/sof/trace/trace.h]: outdated

File: src/trace/trace.c (hunk 27 above)
Comment (ktrzcinx) [reply src/trace/trace.c]: assertion replaced with log error and return NULL

**Comment by ktrzcinx** [comment]
I moved FW part to separate PR, to make this one little shorter. Both PR may be merged independently.

File: tools/logger/logger.c [hunk 32]
```diff
@@ -140,9 +142,39 @@ static int configure_uart(const char *file, unsigned int baud)
...
 }
 
+/* Concantenate `config->filter_config` with `input` + `\n` */
+static int append_filter_config(struct convert_config *config,
+				const char *input)
+{
+	char *buff;
+	size_t len;
+
+	if (!config->filter_config) {
```
Comment (lgirdwood) [inline tools/logger/logger.c:152]: Can you add a comment each block here, I'm trying to understand why we cant use sprintf() here.

File: tools/logger/logger.c [hunk 33]
```diff
@@ -140,9 +142,39 @@ static int configure_uart(const char *file, unsigned int baud)
...
+		config->filter_config_capacity = 128;
```
Comment (lgirdwood) [inline tools/logger/logger.c:153]: magic number

File: tools/logger/filter.c [hunk 34]
```diff
@@ -0,0 +1,330 @@
...
+
+struct filter_element {
+	struct list_item list;
+	int32_t uuid_id;	/**< type id, or -1 when not important */
+	int32_t comp_id;	/**< component id or -1 when not important */
+	int32_t pipe_id;	/**< pipeline id or -1 when not important */
+	int32_t log_level;	/**< new log level value */
+};
+
+static char *ltrim(char *s)
```
Comment (lgirdwood) [inline tools/logger/filter.c:50]: inline comment. wht does this do ?

File: tools/logger/filter.c [hunk 35]
```diff
@@ -0,0 +1,330 @@
...
+{
+	while (isspace(*s))
+		s++;
+	return s;
+}
+
+static char *rtrim(char *s)
```
Comment (lgirdwood) [inline tools/logger/filter.c:57]: ditto

File: tools/logger/filter.c [hunk 36]
```diff
@@ -0,0 +1,330 @@
...
+	char *ptr = s + strlen(s) - 1;
+
+	while (ptr >= s && isspace(*ptr)) {
+		*ptr = '\0';
+		--ptr;
+	}
+	return s;
+}
+
+static char *trim(char *s)
```
Comment (lgirdwood) [inline tools/logger/filter.c:68]: ditto

File: tools/logger/filter.c [hunk 37]
```diff
@@ -0,0 +1,330 @@
...
+			return log_level_dict[i].log_level;
+	}
+	return -1;
+}
+
+static char *filter_parse_component_name(const struct snd_sof_uids_header *uids_dict,
+					 char *input_str,
+					 struct filter_element *out)
+{
+	static char scan_format_string[32] = "";
```
Comment (lgirdwood) [inline tools/logger/filter.c:116]: Is 32 bytes for a single component or command line ? This should also be a macro.

**Review by lgirdwood — CHANGES_REQUESTED** [review]
Excellent work ! @mmaka1 do we need a debug ABI update here ? I guess we do ?

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
Jenkins CI known issue and unrelated to CI.

File: tools/logger/filter.c [hunk 38]
```diff
@@ -0,0 +1,333 @@
...
+
+	list_init(&filter_list);
+
+	/* parse `input_str` line by line */
+	line_end = strchr(input_str, '\n');
+	while (line_end) {
+		line_end[0] = '\0';
+		ret = filter_parse_entry(uids_dict, input_str, &filter_list);
+		if (ret < 0)
+			goto err;
```
Comment (marc-hb) [inline tools/logger/filter.c:299]: As reported by gcc 9.3 -Werr, this leaved out_fd uninitialized and can make fclose() below trying to close a random file descriptor. Fixed by #3412 

By the way are we still stuck with antique start-of-block declarations because of some ancient xtensa toolchain or can we starting using C99's safer variable declarations?

Comment (lgirdwood) [reply tools/logger/filter.c:299]: @marc-hb the start of block declaration is part of the "safety critical" coding standard. I would suggest speaking to a safety critical expert and see what their take is on C99 declarations. I think last time that @plbossart or I checked this was not yet approved.
